Add colourful tree indicators for comments in files

GregSherman opened the pull request:
<img width="207" alt="Screenshot 2023-10-23 at 1 48 56 AM" src="https://github.com/mozilla/assay/assets/63402349/03e163f3-dc9f-43d2-a036-a9afc089388c">

The file tree now shows which files/folders have comments in them. To do this, I had to rework the cache to have layered keys depending on the path parts (so i could propagate upwards and quickly check if the folder has subcontents with comments). 

The cache also has a new function to remove empty keys. Previously, I just had "deleted" comments be an empty string, then ignored the empty string. Now, if a key has no value, that key is removed from the object. Then we check its parent to see if it still has values. This is much more like a proper cache.

the tree decorators are updated on load of the extension, when a new comment is made, when a comment is deleted, or when text editors are swapped.

The decorators are also quite easy to extend. When linting becomes a thing, we can have multiple decorators (i think up to 2 characters at a time) on a single file.

diox commented:
Weird issue I saw while playing with this locally: Initially, it was showing me a `C` indicator for the whole add-on, but not in any files in any versions. Exporting comments for either version I had loaded showed nothing, I have no idea why assay showed me an indicator.
![Screenshot from 2023-10-23 14-29-01](https://github.com/mozilla/assay/assets/187006/b4a59ec0-e040-406a-843b-4ba134f96d49)

Also, off-topic about line-indicators we might want to track somewhere if that's not already known:
- Later I added a comment in a file and that went ok, but then after reloading the line indicator icon wasn't displayed until I exported the comments.
- The fact that you can't click on a line indicator icon is super annoying.

GregSherman commented:
> Later I added a comment in a file and that went ok, but then after reloading the line indicator icon wasn't displayed until I exported the comments.

Ah, can't believe I missed that. It's because the comments are only loaded when changing editors. I made it load the comments on reload as well.

> The fact that you can't click on a line indicator icon is super annoying.

Very much agree, unfortunately that is not exposed to us yet. https://github.com/microsoft/vscode/issues/5455 . I added the add/edit comment to the command palette to quickly add a comment instead of having to bring up the context menu every time.

> Initially, it was showing me a C indicator for the whole add-on, but not in any files in any versions.

Weird, I thought I had fixed that and can't seem to reproduce. Have you added comments to this guid before this patch? It might be worth it to ignore the guid from highlighting altogether, since it's a special case anyways (it is not a key in a cache file but rather the name of the cache file)

diox commented:
> Weird, I thought I had fixed that and can't seem to reproduce. Have you added comments to this guid before this patch? It might be worth it to ignore the guid from highlighting altogether, since it's a special case anyways (it is not a key in a cache file but rather the name of the cache file)

I did add comments to that guid at some point in the past, though I don't remember if I still had any when I tried this patch. I tried exporting them and there weren't any. I agree removing the decorations on the guid seem like a good plan.

diox commented on src/views/fileDecorations.ts:
```diff
@@ -0,0 +1,81 @@
+
+  // if there no keys, then we are at the guid path. return false
+  if (keys.length === 0) {
+    return false;
```
:+1: but add a test for that

GregSherman commented on src/views/fileDecorations.ts:
```diff
@@ -0,0 +1,76 @@
+      (fs.lstatSync(uri.fsPath).isDirectory() && (await folderHasComment(uri)))
+    ) {
+      return {
+        badge: "C",
```
![Screenshot 2023-10-24 at 9 32 56 AM](https://github.com/mozilla/assay/assets/63402349/7d27a093-3fbf-4888-81d0-fabaf2603c4b)
Is this the intention?

diox replied:
mmm it worked better on my machine :) I guess you don't have the font for it, so it's probably not safe to use. We should look into that eventually, maybe there is a safer emoji we could use, but feel free to ignore for now.

diox replied:
Here is how it looked like on my machine FWIW 
![Screenshot from 2023-10-24 16-45-03](https://github.com/mozilla/assay/assets/187006/3f0377ee-401c-49f1-a2a0-9148cd0c2c16)
